common/lessonplan/generate.ts

```diff
@@ -0,0 +1,298 @@
+                    }
+
+                    return {
+                        name: file.originalname,
+                        content: content,
+                        type: file.mimetype,
+                    };
+                } catch (error) {
+                    logger.error(`Error fetching or parsing file with UUID ${uuid}: ${error.message}`);
+                    throw error;
+                }
+            })
+        );
+
+        const validFiles = allFiles.filter((file): file is NonNullable<typeof file> => file !== null);
+
+        if (emptyFiles.length > 0) {
+            throw new Error(`The following files are empty or their content could not be processed: ${emptyFiles.join(', ')}`);
+        }
+
+        // Combine all file contents
+        combinedContent = validFiles.map((file) => file.content).join('\n\n');
+    }
+
+    // Check for OpenAI API key
+    if (!process.env.OPENAI_API_KEY) {
+        throw new Error('OpenAI API key is not set. Please set the OPENAI_API_KEY environment variable.');
+    }
+
+    // Initialize the ChatOpenAI model
+    const model = new ChatOpenAI({
+        modelName: MODEL_NAME,
+        temperature: 0,
+    });
+
+    // Create a new plan schema based on expected outputs
+    const dynamicPlan = z.object(Object.fromEntries(Object.entries(plan.shape).filter(([key]) => !expectedOutputs || expectedOutputs.includes(key))));
+
+    const structuredLlm = model.withStructuredOutput(dynamicPlan);
+
+    try {
+        logger.info('Generating lesson plan...');
+        const requestedFields =
+            expectedOutputs && expectedOutputs.length > 0
+                ? expectedOutputs.map((field) => `${field}: {${field}}`).join('\n')
+                : Object.keys(plan.shape)
+                      .map((field) => `${field}: {${field}}`)
+                      .join('\n');
+
+        let finalPrompt = `${LESSON_PLAN_PROMPT.replace('{requestedFields}', requestedFields).replace(
+            '{language}',
+            language
+        )}\n\nCreate a lesson plan for ${subject} students in grade ${grade} in the state ${state}, for a ${schoolType} school. The lesson should last ${duration} minutes. ${prompt}`;
+
+        if (combinedContent) {
+            finalPrompt += `\n\nInclude relevant content from the provided materials:\n${combinedContent}`;
+        }
+
+        // Only count tokens if the character count exceeds MAX_TOKENS
+        if (finalPrompt.length > MAX_TOKENS) {
+            const tokenCount = countTokens(finalPrompt, MODEL_NAME);
+            if (tokenCount > MAX_TOKENS) {
+                logger.warn(`Prompt exceeds token limit. Prompt has ${tokenCount} tokens but the maximum allowed is ${MAX_TOKENS} tokens. Truncating...`);
+                const truncatedContent = truncateText(
+                    combinedContent,
+                    MODEL_NAME,
+                    MAX_TOKENS - countTokens(finalPrompt.replace(combinedContent, ''), MODEL_NAME)
```

Comment:
if `MAX_TOKENS - countTokens(finalPrompt.replace(combinedContent, ''), MODEL_NAME) > MAX_TOKENS` the provided prompt might still be too long? Then we just fail with the OpenAI request?